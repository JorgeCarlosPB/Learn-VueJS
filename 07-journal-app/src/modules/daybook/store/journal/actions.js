
//Actions son acciones asincronas para llamar mutaciÃ³n
// export const myAction = async ({commit}) =>{

import journalApi from "@/api/journalApi"

    
// }

export const loadEntries = async ({commit}) =>{
    const {data} = await journalApi.get('/entries.json')
    if( !data){
        commit('setEntries', [])
        return
    } 
    const entries = []
    for (let id of Object.keys(data)){
        entries.push({
            id,
            ...data[id]
        })
    }

    commit('setEntries', entries)
}

export const updateEntry = async ({commit}, entry) => {                //entry debe de ser un parametro

    const {date, picture, text} = entry
    const dataToSave = {date, picture, text}

    const resp=await journalApi.put(`/entries/${entry.id}.json`,dataToSave)
    dataToSave.id = entry.id
    //Extraer solo lo que se necesita  //-id

    //await jounrlaApi.put(PATH .json, dataToSave)

    //Commit de una mutation -> updateEntry
    commit('updateEntry',{...dataToSave})

}

export const createEntry = async ({commit}, entry) => {
    //dataToSave
    const {date,picture, text} = entry
    const dataToSave = {date,picture, text}

    const {data} = await journalApi.post(`entries.json`,dataToSave)
    //cons {data} = await journalApi.post(PATH.json, dataToSave)

    dataToSave.id = data.name
    commit('addEntry', dataToSave)
    //data = {"name": "MarijsKGESa-SGsef"}

    return data.name
}

export const deleteEntry = async ({commit},id)=>{

    //Await jorunalApi.delete(path)
    await journalApi.delete(`/entries/${id}.json`)

    commit('deleteEntry',id)

    return id

}