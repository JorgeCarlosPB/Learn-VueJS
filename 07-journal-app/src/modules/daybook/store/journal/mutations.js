//son sincronas y hacen modificaciones del state
// export const myAction = {state} =>{
    
// }

export const setEntries = (state, entries) =>{
    state.entries = [...state.entries, ...entries]
    state.isLoading = false
    
}

export const updateEntry = (state, entry) =>{  //entry actualizada

    //state.entries => un arreglo...
    const idx = state.entries.map(e => e.id).indexOf(entry.id)
    state.entries[idx]=entry

    //state.entries => 
    
}

export const addEntry = (state, entry) =>{
    //state ->entries -> la nueva entrada debe de ser la primera

    state.entries = [entry, ...state.entries]
}

export const deleteEntry = (state,id) => {
    //remover del state.entries borrar la entrada por ese ID

    state.entries = state.entries.filter(entry => entry.id !== id)
}