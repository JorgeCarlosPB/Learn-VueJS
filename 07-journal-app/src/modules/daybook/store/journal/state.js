
//el state es reactivo y cuando cambie se notifica a todos los lugres que estÃ¨n escucando el state
// export default () =>({
    
// })

export default()=>({
    isLoading: true,
    entries: []

})