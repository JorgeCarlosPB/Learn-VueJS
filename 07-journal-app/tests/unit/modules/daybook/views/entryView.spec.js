import { shallowMount } from '@vue/test-utils'
import {createStore} from 'vuex'

import Swal from 'sweetalert2'
import journal from '@/modules/daybook/store/journal'
import {journalState} from '../../../mock-data/test-journal-state'
import EntryView from '@/modules/daybook/views/EntryView.vue'

const createVuexStore = (initialState) =>
    createStore({
        modules: {
            journal: {
                ...journal,
                state: {...initialState}
            }
        }
    })

    jest.mock('sweetalert2',() =>({
        fire: jest.fn(),
        showLoading: jest.fn(),
        close: jest.fn()
    }))

describe('Pruebas en el entryview', ()=>{

    const store = createVuexStore(journalState)
    store.dispatch = jest.fn()

    const mockRouter = {
        push: jest.fn()
    }

    let wrapper 
    beforeEach(()=>{
        jest.clearAllMocks()
        wrapper = shallowMount(EntryView, {
            props: {
                id: '-N5H2wq9F6G2tpFcl1PC'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [store]
            }
        })
    })

    test('Debe de sacar al usuario porque el id no existe', ()=>{

        const wrapper = shallowMount(EntryView, {
            props: {
                id: 'Este ID no existe en el store'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [store]
            }
        })

        expect(mockRouter.push).toHaveBeenCalledWith({name: 'no-entry'})
    })

    test('Debe de mostrar la entrada correctamente', ()=>{

        expect(wrapper.html()).toMatchSnapshot()
        expect(mockRouter.push).not.toHaveBeenCalled()

    
    })

    test('Debe de borrar la entrada y salir', (done)=>{

        Swal.fire.mockReturnValueOnce( Promise.resolve({isConfirmed: true}))
        wrapper.find('.btn-danger').trigger('click')

        expect(Swal.fire).toHaveBeenCalledWith({
            title: '¿Está seguro?',
            text: 'Una vez borrado, no se puede recuperar',
            showDenyButton: true,
            confirmButtonText: 'Sí, estoy seguro',
        }) 

        setTimeout(()=>{
            expect(store.dispatch).toHaveBeenCalledWith("journal/deleteEntry", "-N5H2wq9F6G2tpFcl1PC")
            expect(mockRouter.push).toHaveBeenCalled()
            done()
        },1)

        

    
    })

    
    
})