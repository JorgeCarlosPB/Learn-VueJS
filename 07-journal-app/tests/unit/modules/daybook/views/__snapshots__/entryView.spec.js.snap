// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el entryview Debe de mostrar la entrada correctamente 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-3 fw-bold">23</span><span class="mx-1 fs-3">Junio</span><span class="mx-2 fs-4 fw-light">2022, Jueves</span></div>
  <div><input type="file" style="display: none;"><button class="btn btn-danger mx-2"> Borrar <i class="fa fa-trash-alt"></i></button><button class="btn btn-primary"> Subir foto <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="¿Qué sucedió hoy?">
            </textarea></div>
<!--v-if-->
<!--v-if-->
<my-fab-stub icon="fa-save"></my-fab-stub>
`;
